### PR TITLE
(Fix #676) Fix and enhance openssl compatibility

### LINKED ISSUE #676: Compiler warning under OpenSSL 1.1.1 / FreeBSD 12 / clang 6.0.1 and more
```
In file included from tls.c:33:
/usr/local/include/openssl/x509v3.h:634:10: warning: 'hex_to_string' macro redefined [-Wmacro-redefined]
# define hex_to_string OPENSSL_buf2hexstr
         ^
../config.h:494:9: note: previous definition is here
#define hex_to_string OPENSSL_hexstr2buf
        ^
In file included from tls.c:33:
/usr/local/include/openssl/x509v3.h:635:10: warning: 'string_to_hex' macro redefined [-Wmacro-redefined]
# define string_to_hex OPENSSL_hexstr2buf
         ^
../config.h:513:9: note: previous definition is here
#define string_to_hex OPENSSL_buf2hexstr
        ^
2 warnings generated.
```

No big deal, but:

Using those functions was hacky from the beginning.

Later we had to workaround API changes.

So now its time to finally fix our code.

However, a simple fix could be to not use the old function names and define the old ones to the new ones if need be, but to use the new function names and define the new ones to the old ones if need be.

Better we don't use the old "interal" functions hex_to_string() and string_to_hex() at all, but bring our own compat functions, like irssi does with src/core/misc.c:binary_to_hex().

While at it:

I miss proper error handling in tls.c: `X509_digest(cert, EVP_md5(), md, &len); /* MD5 hash */`, we should test for return value 0/1 and maybe not for `len <= sizeof(md)` and as i already said elsewhere, i would like to see EVP_sha256 or something here, but not md5 or sha1, maybe as fallback. Should be easy to fix, because the functionality is only used to print out a fingerprint to human.

**While analyzing and fixing, I just found another bug:**
openssl-1.1.1:x509v3.h:
```
# define hex_to_string OPENSSL_buf2hexstr
# define string_to_hex OPENSSL_hexstr2buf
```
eggdrop:aclocal.m4:
```
AC_DEFINE([hex_to_string], [OPENSSL_hexstr2buf]
AC_DEFINE([string_to_hex], [OPENSSL_buf2hexstr]
```
eggdrop:config.h:
```
#define hex_to_string OPENSSL_hexstr2buf
#define string_to_hex OPENSSL_buf2hexstr
```
Maybe the bug in aclocal.m4 and config.h compensate each other. Someone really wanted to confuse me. But don't worry, i got this one.

### PR DESCRIPTION
**Pull request by michaelortmann**
Found by: michaelortmann
Patch by: michaelortmann
Fixes: #676

One-line summary:
Fix and enhance openssl compatibility

Additional description (if needed):
1. We were using old functions and if they were not available we defined them to use the new ones. This patch reverses this logic and automagically gets rid of the compiler warning.
2. We were printing MD5 and SHA1 fingerprints. This patch changes this, and we ill print SHA-256 and SHA1.
4. I also enhanced the error handling of function X509_digest().
5. Last but not least, i enhanced variable definition `unsigned i` -> `unsigned int i` and other variable definitions.
6. Fixed a well hidden bug. eggdrops aclocal.m4 /config.h and tls.c linked and used the wrong functions. Example: hex_to_string is OPENSSL_buf2hexstr and not OPENSSL_hexstr2buf. Eggdrop didn't fail, but there may be consequential errors.

**Please run misc/runautotools after merge**

Test cases demonstrating functionality (if applicable):

**Test 1: compile unter FreeBSD 12.0-ALPHA8 / clang 6.0.1, where i initially saw the compiler warning with installed openssl 1.1.1:**
`$ pkg install openssl111`

before:
```
$ ./configure --with-sslinc=/usr/local/include --with-ssllib=/usr/local/lib
[...]
checking for hex_to_string... no
checking for OPENSSL_hexstr2buf... yes
checking for string_to_hex... no
checking for OPENSSL_buf2hexstr... yes
[...]
$ make config
[...]
$make
[...]
Making core eggdrop files...

cc -g -O2 -pipe -Wall -I.. -I.. -I/usr/local/include -DHAVE_CONFIG_H -I/usr/local/include/tcl8.7 -g3 -DDEBUG -DDEBUG_ASSERT -DDEBUG_MEM -DDEBUG_DNS  -c tls.c
In file included from tls.c:33:
/usr/local/include/openssl/x509v3.h:634:10: warning: 'hex_to_string' macro redefined [-Wmacro-redefined]
# define hex_to_string OPENSSL_buf2hexstr
         ^
../config.h:494:9: note: previous definition is here
#define hex_to_string OPENSSL_hexstr2buf
        ^
In file included from tls.c:33:
/usr/local/include/openssl/x509v3.h:635:10: warning: 'string_to_hex' macro redefined [-Wmacro-redefined]
# define string_to_hex OPENSSL_hexstr2buf
         ^
../config.h:513:9: note: previous definition is here
#define string_to_hex OPENSSL_buf2hexstr
        ^
2 warnings generated.
[...]
```

after:
```
$ ./configure --with-sslinc=/usr/local/include --with-ssllib=/usr/local/lib
[...]
checking for OPENSSL_buf2hexstr... yes
checking for OPENSSL_hexstr2buf... yes
[...]
SSL/TLS Support: yes (OpenSSL 1.1.1  11 Sep 2018)
[...]
$ make config
[...]
$make
[...]
```

OK. Note: There is not one single compiler warning under this setup, yeah!

**Test 2: compile unter FreeBSD 12.0-ALPHA8 / clang 6.0.1, where i initially saw the compiler warning with preinstalled openssl 1.0.2p:**
`$ pkg remove openssl111`

```
$ ./configure
[...]
checking for OPENSSL_buf2hexstr... no
checking for hex_to_string... yes
checking for OPENSSL_hexstr2buf... no
checking for string_to_hex... yes
[...]
SSL/TLS Support: yes (OpenSSL 1.0.2p-freebsd  14 Aug 2018)
[...]
$ make config
[...]
$make
[...]
```

OK. Note: There is not one single compiler warning under this setup, yeah!

**Test 3: link share bots under Linux 4.18.10 / gcc 8.2.1with installed openssl 1.1.1:**
```
.link sslbot
[01:35:33] #-HQ# link sslbot
[01:35:33] Linking to sslbot at zen:2514 ...
[01:35:34] TLS: handshake successful. Secure connection established.
[01:35:34] TLS: certificate subject: C=EU, O=Eggheads, OU=Botnet, UID=sslbot
[01:35:34] TLS: certificate issuer: C=EU, O=Eggheads, OU=Botnet, UID=sslbot
[01:35:34] TLS: certificate SHA-256 Fingerprint: 31:96:24:A1:E9:90:0F:64:D9:BC:CC:92:D4:47:83:96:80:CE:FB:7F:95:84:46:77:4B:49:48:47:12:31:5F:5B
[01:35:34] TLS: certificate SHA1 Fingerprint: 09:A0:58:11:CB:2C:40:5E:61:DA:A2:F9:55:CB:D4:01:2E:35:5C:43
[01:35:34] TLS: certificate valid from Oct  4 02:10:32 2018 GMT to Oct  4 02:10:32 2019 GMT
[01:35:34] TLS: cipher used: TLS_AES_256_GCM_SHA384 TLSv1.3; 256 bits (256 secret)
[01:35:34] TLS: handshake successful. Secure connection established.
[01:35:34] TLS: certificate subject: C=EU, O=Eggheads, OU=Botnet, UID=sslbot
[01:35:34] TLS: certificate issuer: C=EU, O=Eggheads, OU=Botnet, UID=sslbot
[01:35:34] TLS: certificate SHA-256 Fingerprint: 31:96:24:A1:E9:90:0F:64:D9:BC:CC:92:D4:47:83:96:80:CE:FB:7F:95:84:46:77:4B:49:48:47:12:31:5F:5B
[01:35:34] TLS: certificate SHA1 Fingerprint: 09:A0:58:11:CB:2C:40:5E:61:DA:A2:F9:55:CB:D4:01:2E:35:5C:43
[01:35:34] TLS: certificate valid from Oct  4 02:10:32 2018 GMT to Oct  4 02:10:32 2019 GMT
[01:35:34] TLS: cipher used: TLS_AES_256_GCM_SHA384 TLSv1.3; 256 bits (256 secret)
[01:35:35] TLS: handshake successful. Secure connection established.
[01:35:35] TLS: certificate subject: C=EU, O=Eggheads, OU=Botnet, UID=sslbot
[01:35:35] TLS: certificate issuer: C=EU, O=Eggheads, OU=Botnet, UID=sslbot
[01:35:35] TLS: certificate SHA-256 Fingerprint: 31:96:24:A1:E9:90:0F:64:D9:BC:CC:92:D4:47:83:96:80:CE:FB:7F:95:84:46:77:4B:49:48:47:12:31:5F:5B
[01:35:35] TLS: certificate SHA1 Fingerprint: 09:A0:58:11:CB:2C:40:5E:61:DA:A2:F9:55:CB:D4:01:2E:35:5C:43
[01:35:35] TLS: certificate valid from Oct  4 02:10:32 2018 GMT to Oct  4 02:10:32 2019 GMT
[01:35:35] TLS: cipher used: TLS_AES_256_GCM_SHA384 TLSv1.3; 256 bits (256 secret)
[01:35:35] Received challenge from sslbot... sending response ...
[01:35:35] Linked to sslbot.
[01:35:35] Creating resync buffer for sslbot
[01:35:35] Sending user file send request to sslbot
[01:35:36] TLS: handshake successful. Secure connection established.
[01:35:36] TLS: certificate subject: C=EU, O=Eggheads, OU=Botnet, UID=sslbot
[01:35:36] TLS: certificate issuer: C=EU, O=Eggheads, OU=Botnet, UID=sslbot
[01:35:36] TLS: certificate SHA-256 Fingerprint: 31:96:24:A1:E9:90:0F:64:D9:BC:CC:92:D4:47:83:96:80:CE:FB:7F:95:84:46:77:4B:49:48:47:12:31:5F:5B
[01:35:36] TLS: certificate SHA1 Fingerprint: 09:A0:58:11:CB:2C:40:5E:61:DA:A2:F9:55:CB:D4:01:2E:35:5C:43
[01:35:36] TLS: certificate valid from Oct  4 02:10:32 2018 GMT to Oct  4 02:10:32 2019 GMT
[01:35:36] TLS: cipher used: TLS_AES_256_GCM_SHA384 TLSv1.3; 256 bits (256 secret)
[01:35:36] TLS: handshake successful. Secure connection established.
[01:35:36] TLS: certificate subject: C=EU, O=Eggheads, OU=Botnet, UID=sslbot
[01:35:36] TLS: certificate issuer: C=EU, O=Eggheads, OU=Botnet, UID=sslbot
[01:35:36] TLS: certificate SHA-256 Fingerprint: 31:96:24:A1:E9:90:0F:64:D9:BC:CC:92:D4:47:83:96:80:CE:FB:7F:95:84:46:77:4B:49:48:47:12:31:5F:5B
[01:35:36] TLS: certificate SHA1 Fingerprint: 09:A0:58:11:CB:2C:40:5E:61:DA:A2:F9:55:CB:D4:01:2E:35:5C:43
[01:35:36] TLS: certificate valid from Oct  4 02:10:32 2018 GMT to Oct  4 02:10:32 2019 GMT
[01:35:36] TLS: cipher used: TLS_AES_256_GCM_SHA384 TLSv1.3; 256 bits (256 secret)
[01:35:36] TLS: handshake successful. Secure connection established.
[01:35:36] TLS: certificate subject: C=EU, O=Eggheads, OU=Botnet, UID=sslbot
[01:35:36] TLS: certificate issuer: C=EU, O=Eggheads, OU=Botnet, UID=sslbot
[01:35:36] TLS: certificate SHA-256 Fingerprint: 31:96:24:A1:E9:90:0F:64:D9:BC:CC:92:D4:47:83:96:80:CE:FB:7F:95:84:46:77:4B:49:48:47:12:31:5F:5B
[01:35:36] TLS: certificate SHA1 Fingerprint: 09:A0:58:11:CB:2C:40:5E:61:DA:A2:F9:55:CB:D4:01:2E:35:5C:43
[01:35:36] TLS: certificate valid from Oct  4 02:10:32 2018 GMT to Oct  4 02:10:32 2019 GMT
[01:35:36] TLS: cipher used: TLS_AES_256_GCM_SHA384 TLSv1.3; 256 bits (256 secret)
[01:35:37] Completed userfile transfer to sslbot.
```

OK. Note: Fingerprints are now (SHA-256 and SHA1) instead of (MD5 and SHA1).

**Test 4: compile unter OpenBSD 6.3 / gcc 4.2.1 with libressl 2.7.2:**
```
$ ./configure
[...]
checking for OPENSSL_buf2hexstr... no
checking for hex_to_string... yes
checking for OPENSSL_hexstr2buf... no
checking for string_to_hex... yes
[...]
SSL/TLS Support: yes (LibreSSL 2.7.2)
[...]
$ make config
[...]
$make
[...]
```

OK.

### CODE REVIEW

**Review by thommey — APPROVED** [review]

**Comment by vanosg** [comment]
build

File: src/tls.c [hunk 1]
```diff
@@ -474,12 +474,12 @@ static char *ssl_printnum(ASN1_INTEGER *i)
...
  */
-static void ssl_showcert(X509 *cert, int loglev)
+static void ssl_showcert(X509 *cert, const int loglev)
 {
-  char *buf, *from, *to;
   X509_NAME *name;
-  unsigned int len;
+  char *buf, *from, *to;
   unsigned char md[EVP_MAX_MD_SIZE];
+  unsigned int len;
```
Comment (Cizzle) [inline src/tls.c]: Why the reordering?

File: src/tls.c [hunk 2]
```diff
@@ -496,15 +496,13 @@ static void ssl_showcert(X509 *cert, int loglev)
     putlog(loglev, "*", "TLS: cannot get issuer name from certificate!");
 
   /* Fingerprints */
-  X509_digest(cert, EVP_md5(), md, &len); /* MD5 hash */
-  if (len <= sizeof(md)) {
-    buf = hex_to_string(md, len);
-    putlog(loglev, "*", "TLS: certificate MD5 Fingerprint: %s", buf);
+  if(X509_digest(cert, EVP_sha256(), md, &len)) {
```
Comment (Cizzle) [inline src/tls.c]: Add space between "if" and "(". Is the `len <= sizeof(md)` not needed anymore? Same for below.

Comment (michaelortmann) [reply src/tls.c]: exactly. the len <= sizeof looks like a test, if the function was successful and was replaced with a _real_ test of success. its like main(argv, argc)... you dont test if the argc count is right here. same with len for md.

File: src/tls.c (hunk 1 above)
Comment (michaelortmann) [reply src/tls.c]: i undid the reordering that slipped in during editing.

**Comment by Cizzle** [comment]
LGTM

**Comment by vanosg** [comment]
build

**Pull request merged**